include/Gaffer/NameValuePlug.h

```diff
@@ -0,0 +1,123 @@
+//  met:
+//
+//      * Redistributions of source code must retain the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer.
+//
+//      * Redistributions in binary form must reproduce the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#ifndef GAFFER_NAMEVALUEPLUG_H
+#define GAFFER_NAMEVALUEPLUG_H
+
+#include "Gaffer/TypeIds.h"
+
+#include "Gaffer/TypedPlug.h"
+#include "Gaffer/StringPlug.h"
+
+namespace Gaffer
+{
+
+class GAFFER_API NameValuePlug : public Gaffer::ValuePlug
+{
+
+	public :
+
+		IE_CORE_DECLARERUNTIMETYPEDEXTENSION( Gaffer::NameValuePlug, NameValuePlugTypeId, Gaffer::ValuePlug );
+
+		// Bare constructor required for compatibility with old CompoundDataPlug::MemberPlug constructor.
+		// Deprecated, and dangerous, since if you don't manually construct child plugs in the expected order of
+		// "name", "value", and optionally "enabled" then you will get a crash.
+		NameValuePlug(
+			const std::string &name=defaultName<NameValuePlug>(),
+			Direction direction=In,
+			unsigned flags=Default
+		);
+
+		// Construct a NameValuePlug with the "name" and "value" children.  The value plug
+		// can be constructed either based on a IECore::Data default value, or by supplying
+		// a plug.  In the variant which takes a ValuePlug, NameValuePlug will take ownership
+		// of this plug ( so be careful about using it afterwards ).
+		NameValuePlug(
+			const std::string &nameDefault,
+			const IECore::Data *valueDefault,
+			const std::string &name=defaultName<NameValuePlug>(),
+			Direction direction=In,
+			unsigned flags=Default
+		);
+
+		NameValuePlug(
+			const std::string &nameDefault,
+			Gaffer::PlugPtr valuePlug,
```

Comment:
Is it intentional that this takes a `Plug` rather than a `ValuePlug`? I had assumed it was perhaps to allow the use of `ClosurePlug` as a value, but since `NameValuePlug` derives from `ValuePlug`, it requires all children to be `ValuePlugs` anyway. Taking any old `Plug` is good from my point of view (it'll be useful for NameSwitch I think, as well as OSLObject), but we should be deriving from Plug to do that.